be/src/formats/parquet/file_writer.cpp

```diff
@@ -177,136 +191,570 @@ void ParquetBuildHelper::build_compression_type(::parquet::WriterProperties::Bui
+        if (idx == Context::kNULL || (nulls != nullptr && nulls[idx])) {
+            write(def_level, rep_level, -1);
+            continue;
+        }
+        // TODO: add util function
+        int32_t unix_days = raw_col[idx]._julian - DateValue::create(1970, 1, 1)._julian;
+        write(def_level + node->is_optional(), rep_level, unix_days);
+    }
+
+    _buffered_values_estimate[_col_idx] = col_writer->EstimatedBufferedValueBytes();
+    _col_idx++;
+    return Status::OK();
+}
+
+Status FileWriterBase::_add_datetime_column_chunk(const Context& parent_ctx, const TypeDescriptor& type_desc,
+                                                  ::parquet::schema::NodePtr node, ColumnPtr col) {
+    unsigned char* nulls = nullptr;
+    if (col->is_nullable()) {
+        const auto null_column = down_cast<NullableColumn*>(col.get())->null_column();
+        nulls = null_column->get_data().data();
     }
 
+    const auto data_column = ColumnHelper::get_data_column(col.get());
+    auto raw_col = down_cast<const RunTimeColumnType<TYPE_DATETIME>*>(data_column)->get_data().data();
+
+    _generate_rg_writer();
+    auto col_writer = down_cast<::parquet::Int64Writer*>(_rg_writer->column(_col_idx));
+    DCHECK(col_writer != nullptr);
+
+    auto write = [&](int16_t def_level, int16_t rep_level, int64_t value) {
+        col_writer->WriteBatch(1, &def_level, &rep_level, &value);
+    };
+
+    for (auto i = 0; i < parent_ctx.size(); i++) {
+        auto [idx, def_level, rep_level] = parent_ctx.get(i);
+        if (idx == Context::kNULL || (nulls != nullptr && nulls[idx])) {
+            write(def_level, rep_level, -1);
+            continue;
+        }
+        write(def_level + node->is_optional(), rep_level,
+              raw_col[idx].to_unix_second() * 1000); // seconds -> milliseconds
+    }
+
+    _buffered_values_estimate[_col_idx] = col_writer->EstimatedBufferedValueBytes();
+    _col_idx++;
+    return Status::OK();
+}
+
+template <LogicalType lt, ::parquet::Type::type pt>
+Status FileWriterBase::_add_int_column_chunk(const Context& parent_ctx, const TypeDescriptor& type_desc,
+                                             ::parquet::schema::NodePtr node, ColumnPtr col) {
+    unsigned char* nulls = nullptr;
+    if (col->is_nullable()) {
+        const auto null_column = down_cast<NullableColumn*>(col.get())->null_column();
+        nulls = null_column->get_data().data();
+    }
+
+    const auto data_column = ColumnHelper::get_data_column(col.get());
+    const auto raw_col = down_cast<RunTimeColumnType<lt>*>(data_column)->get_data().data();
+
+    _generate_rg_writer();
+    auto col_writer =
+            down_cast<::parquet::TypedColumnWriter<::parquet::PhysicalType<pt>>*>(_rg_writer->column(_col_idx));
+    DCHECK(col_writer != nullptr);
+
+    auto write = [&](int16_t def_level, int16_t rep_level, typename ::parquet::type_traits<pt>::value_type value) {
+        col_writer->WriteBatch(1, &def_level, &rep_level, &value);
```

Comment:
why only one row

Reply:
Do you want to write one row at a time?

Reply:
There may be null values outside of `raw_col` because of nested types. Thus we cannot write the entire column with one `WriteBatch`.

These values will be buffered before flushing. I think the performance is not an issue.

Reply:
Fixed. Call `WriteBatch` once for every primitive column.